### PR TITLE
use a struct to wrap a valid schema and skip schema validation

### PR DESCRIPTION
**Pull request by lud**
See https://github.com/dashbitco/nimble_options/issues/69

This change makes possible to obtain a valid schema, and then validate options directly without validating the schema itself whenever options are to be validated.

I defined the struct directly in the NimbleOptions module, instead of creating a NimbleOptions.Schema module, because from there, I guess a couple options could be added to the struct, that would be passed into the validation, if someday `validate/2` becomes `validate/3` and support validation options.

(this is also why the flag is called `schema_valid?` and not just `valid?`)

Have a nice day :)

### CODE REVIEW

File: lib/nimble_options.ex [hunk 1]
```diff
@@ -188,9 +188,47 @@ defmodule NimbleOptions do
...
+      ...>     required: true,
+      ...>     type: :string
+      ...>   ]
+      ...> ]
+      ...>
+      ...> schema = NimbleOptions.new!(raw_schema)
+      ...>
+      ...> config = [hostname: "elixir-lang.org"]
+      ...>
+      ...> NimbleOptions.validate(config, schema)
```
Comment (josevalim) [inline lib/nimble_options.ex:216]:
```suggestion
      ...> NimbleOptions.validate([hostname: "elixir-lang.org"], schema)
```

**Comment by lud** [comment]
One note though: The docs function do not work with the `recursive_schema()` fixture from the tests. I did not try to change to much things to make it work, as I guessed the recursive_schema was only to test documenting the `NimbleOptions@options_schema` itself.

File: lib/nimble_options.ex [hunk 2]
```diff
@@ -188,9 +188,47 @@ defmodule NimbleOptions do
...
+      @options_schema NimbleOptions.new!([...])
+
+  This schema will be validated at compile time. Calling `docs/1` on that schema is also
+  supported.
   """
 
   alias NimbleOptions.ValidationError
+  alias __MODULE__
+
+  defstruct schema: [], schema_valid?: false
```
Comment (josevalim) [inline lib/nimble_options.ex:231]: We should probably remove the `schema_valid?` field. I don't think we should ever build a schema that is invalid. :)

File: lib/nimble_options.ex [hunk 3]
```diff
@@ -245,14 +290,32 @@ defmodule NimbleOptions do
...
       {:error, %ValidationError{} = error} -> raise error
     end
   end
 
+  @doc """
+  Validates the given `schema` and returns a wrapped schema to be used with `validate/2`.
+
+  If the given schema is not valid, raises a `NimbleOptions.ValidationError`.
+  """
+  @spec new!(schema()) :: t() | {:error, ValidationError.t()}
```
Comment (josevalim) [inline lib/nimble_options.ex:306]:
```suggestion
  @spec new!(schema()) :: t()
```

Comment (whatyouhide) [inline lib/nimble_options.ex:306]: `{:error, _}` will never be returned here.

```suggestion
  @spec new!(schema()) :: t()
```

File: lib/nimble_options/docs.ex [hunk 4]
```diff
@@ -71,7 +71,7 @@ defmodule NimbleOptions.Docs do
   end
 
   defp get_required_str(schema) do
-    schema[:required] && "Required."
+    (schema[:required] && "Required.") || nil
```
Comment (josevalim) [inline lib/nimble_options/docs.ex:74]:
```suggestion
    if schema[:required], do: "Required."
```

Comment (whatyouhide) [inline lib/nimble_options/docs.ex:74]: Why this change? 🤔

Comment (lud) [reply lib/nimble_options/docs.ex:74]: Hi! `Enum.join(parts, " - ")` returns `"false - ..."` when `schema[:required]` is `false`. This is because the struct contains a schema that was passed through the schema validation function and now contains `false` for undefined docs.

File: lib/nimble_options.ex [hunk 5]
```diff
@@ -188,9 +188,47 @@ defmodule NimbleOptions do
...
+  that function is called. Declaring the schema as a module attribute is supported:
+
+      @options_schema NimbleOptions.new!([...])
+
+  This schema will be validated at compile time. Calling `docs/1` on that schema is also
+  supported.
   """
 
   alias NimbleOptions.ValidationError
+  alias __MODULE__
```
[the remark this answers is not in the document]
Comment (lud) [reply lib/nimble_options.ex:226]: so we can write `%MyModule{}` from `MyModule`. I thought it was necessary, but it does not seems required (anymore?).

File: lib/nimble_options.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (lud) [reply lib/nimble_options.ex:227]: same comment from @josevalim , I will reply here.

This is to support people that would build a litteral struct directly, there is no way to prevent that to my knowledge. But I'm ok to remove the field since both of you seem to agree on that.

Let's go?

File: lib/nimble_options.ex (hunk 5 above)
[the remark this answers is not in the document]
Comment (lud) [reply lib/nimble_options.ex]: i pushed the change

File: lib/nimble_options.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (whatyouhide) [reply lib/nimble_options.ex:227]: Yes let's remove. People will always be able to create a struct with an invalid schema in it.

[the remark this answers is not in the document]
Comment (lud) [reply lib/nimble_options.ex:227]: True. Ok I will do it soon.

Comment (lud) [reply lib/nimble_options.ex]: I pushed the change

File: test/nimble_options_test.exs [hunk 6]
```diff
@@ -1541,6 +1542,46 @@ defmodule NimbleOptionsTest do
...
+        e in FunctionClauseError ->
+          assert Exception.message(e) ==
+                   "no function clause matching in Keyword.has_key?/2"
+
+        e in ArgumentError ->
+          assert Exception.message(e) == """
+                 expected a keyword list, but an entry in the list is not a two-element \
+                 tuple with an atom as its first element, got: {"a_binary_key", []}\
+                 """
+      end
```
Comment (josevalim) [inline test/nimble_options_test.exs:1577]: I'd say we don't care about the type of the error.

```suggestion
      assert catch_error(NimbleOptions.validate([], invalid_struct))
```

File: test/nimble_options_test.exs [hunk 7]
```diff
@@ -1541,6 +1542,46 @@ defmodule NimbleOptionsTest do
...
+      assert_raise FunctionClauseError, fn ->
+        NimbleOptions.new!(invalid_schema)
+      end
+    end
+
+    test "will not be validated once built" do
+      invalid_schema = [{"a_binary_key", []}]
+      invalid_struct = %NimbleOptions{schema: invalid_schema}
+
+      # support elixir 1.6
```
Comment (josevalim) [inline test/nimble_options_test.exs:1563]:
```suggestion
```

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
